rustls/src/crypto/mod.rs

```diff
@@ -149,39 +140,33 @@ pub use crate::msgs::handshake::KeyExchangeAlgorithm;
 /// [provider-example/]: https://github.com/rustls/rustls/tree/main/provider-example/
 /// [rust-crypto]: https://github.com/rustcrypto
 /// [dalek-cryptography]: https://github.com/dalek-cryptography
-pub trait CryptoProvider: SecureRandom + KeyProvider + Send + Sync + Debug + 'static {
-    /// Provide a safe set of cipher suites that can be used as the defaults.
-    ///
-    /// This is used by [`crate::ConfigBuilder::with_safe_defaults()`] and
-    /// [`crate::ConfigBuilder::with_safe_default_cipher_suites()`].
-    ///
-    /// Other (non-default) cipher suites can be provided separately and configured
-    /// by passing them to [`crate::ConfigBuilder::with_cipher_suites()`]. That
-    /// includes cipher suites implemented by a different `CryptoProvider`.
+#[derive(Debug, Clone)]
+pub struct CryptoProvider {
+    /// List of supported ciphersuites, in preference order -- the first element
+    /// is the highest priority.
     ///
     /// The `SupportedCipherSuite` type carries both configuration and implementation.
-    fn default_cipher_suites(&self) -> &'static [suites::SupportedCipherSuite];
+    pub cipher_suites: Vec<suites::SupportedCipherSuite>,
 
-    /// Return a safe set of supported key exchange groups to be used as the defaults.
+    /// List of supported key exchange groups, in preference order -- the
+    /// first element is the highest priority.
     ///
-    /// This is used by [`crate::ConfigBuilder::with_safe_defaults()`] and
-    /// [`crate::ConfigBuilder::with_safe_default_kx_groups()`].
-    ///
-    /// Other (non-default) key exchange groups can be provided separately and configured
-    /// by passing them to [`crate::ConfigBuilder::with_kx_groups()`]. That includes
-    /// key exchange groups implemented by a different `CryptoProvider`.
+    /// The first element in this list is the _default key share algorithm_,
+    /// and in TLS1.3 a key share for it is sent in the client hello.
     ///
     /// The `SupportedKxGroup` type carries both configuration and implementation.
-    fn default_kx_groups(&self) -> &'static [&'static dyn SupportedKxGroup];
+    pub kx_groups: Vec<&'static dyn SupportedKxGroup>,
 
-    /// Return the signature verification algorithms for use with webpki.
+    /// List of signature verification algorithms for use with webpki.
     ///
     /// These are used for both certificate chain verification and handshake signature verification.
-    ///
-    /// This is called by [`crate::ConfigBuilder::with_root_certificates()`],
-    /// [`crate::server::WebPkiClientVerifier::builder_with_provider()`] and
-    /// [`crate::client::WebPkiServerVerifier::builder_with_provider()`].
-    fn signature_verification_algorithms(&self) -> WebPkiSupportedAlgorithms;
+    pub signature_verification_algorithms: WebPkiSupportedAlgorithms,
+
+    /// Source of cryptographically secure random numbers.
+    pub secure_random: &'static dyn SecureRandom,
```

Comment:
Might be nice to experiment with changing this to be a separate commit? That is, I think the change from `CryptoProvider` going from is-a `SecureRandom` to `has-a` `SecureRandom` could I think be orthogonal to `CryptoProvider` going from `trait` to `struct` -- it could return the `&'static dyn SecureRandom` in a method. (Same for `KeyProvider`, of course.)

Reply:
> That is, I think the change from CryptoProvider going from is-a SecureRandom to has-a SecureRandom could I think be orthogonal to CryptoProvider going from trait to struct

Sure, that makes sense :+1: I will split it out.

Reply:
I split out two additional commits for this, one for each new trait.